raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -73,10 +73,11 @@ public int epoch() {
 
     private boolean updateHighWatermark() {
         // Find the largest offset which is replicated to a majority of replicas (the leader counts)
-        ArrayList<FollowerState> followersByFetchOffset = new ArrayList<>(this.followers.values());
+        ArrayList<ReplicaState> followersByFetchOffset = new ArrayList<>(this.voterReplicaStates.values());
         Collections.sort(followersByFetchOffset);
-        int indexOfHw = followers.size() / 2 - (followers.size() % 2 == 0 ? 1 : 0);
+        int indexOfHw = voterReplicaStates.size() / 2;
```

Comment:
Is this a bug fix? For quorum with even numbers (example 4), the index become 2 instead of 1.

Reply:
This was not a bug fix, but a simplification suggested by @guozhangwang. We changed the sort order to be descending so that this logic is equivalent.